orm/iterator_test.go

```diff
@@ -90,48 +94,148 @@ func TestReadAll(t *testing.T) {
+	})
+	tb := tBuilder.Build()
+	ctx := NewMockContext()
+
+	admin := sdk.AccAddress([]byte("admin-address"))
+	g1 := testdata.GroupInfo{
+		Description: "my test 1",
+		Admin:       admin,
+	}
+	g2 := testdata.GroupInfo{
+		Description: "my test 2",
+		Admin:       admin,
+	}
+	g3 := testdata.GroupInfo{
+		Description: "my test 3",
+		Admin:       sdk.AccAddress([]byte("other-admin-address")),
+	}
+	g4 := testdata.GroupInfo{
+		Description: "my test 4",
+		Admin:       admin,
+	}
+	g5 := testdata.GroupInfo{
+		Description: "my test 5",
+		Admin:       sdk.AccAddress([]byte("other-admin-address")),
+	}
+
+	for _, g := range []testdata.GroupInfo{g1, g2, g3, g4, g5} {
+		_, err := tb.Create(ctx, &g)
+		require.NoError(t, err)
+	}
+
+	specs := map[string]struct {
+		pageReq    *query.PageRequest
+		expPageRes *query.PageResponse
+		exp        []testdata.GroupInfo
+		key        []byte
+		expErr     bool
+	}{
+		"one item": {
+			pageReq:    &query.PageRequest{Key: nil, Limit: 1},
+			exp:        []testdata.GroupInfo{g1},
+			expPageRes: &query.PageResponse{Total: 0, NextKey: EncodeSequence(2)},
+			key:        admin,
+		},
+		"with both key and offset": {
+			pageReq: &query.PageRequest{Key: EncodeSequence(2), Offset: 1},
+			expErr:  true,
+			key:     admin,
+		},
+		"up to max": {
+			pageReq:    &query.PageRequest{Key: nil, Limit: 3, CountTotal: true},
+			exp:        []testdata.GroupInfo{g1, g2, g4},
+			expPageRes: &query.PageResponse{Total: 3, NextKey: nil},
+			key:        admin,
+		},
+		"no results": {
+			pageReq:    &query.PageRequest{Key: nil, Limit: 2, CountTotal: true},
+			exp:        []testdata.GroupInfo{},
+			expPageRes: &query.PageResponse{Total: 0, NextKey: nil},
+			key:        sdk.AccAddress([]byte("no-group-address")),
+		},
+		"with offset": {
+			pageReq:    &query.PageRequest{Key: nil, Offset: 1, Limit: 2, CountTotal: true},
+			exp:        []testdata.GroupInfo{g2, g4},
+			expPageRes: &query.PageResponse{Total: 3, NextKey: nil},
+			key:        admin,
+		},
+		"nil/default page req (limit = 100 > number of items)": {
+			pageReq:    nil,
+			exp:        []testdata.GroupInfo{g1, g2, g4},
+			expPageRes: &query.PageResponse{Total: 3, NextKey: nil},
+			key:        admin,
+		},
+		"with key": {
+			pageReq:    &query.PageRequest{Key: EncodeSequence(2), Limit: 10},
+			exp:        []testdata.GroupInfo{g2, g4},
+			expPageRes: &query.PageResponse{Total: 0, NextKey: nil},
+			key:        admin,
+		},
+	}
+	for msg, spec := range specs {
+		t.Run(msg, func(t *testing.T) {
+			var loaded []testdata.GroupInfo
+
+			it, err := idx.GetPaginated(ctx, spec.key, spec.pageReq)
+			require.NoError(t, err)
+
+			res, err := Paginate(it, spec.pageReq, &loaded)
+			if spec.expErr {
+				require.Error(t, err)
+			} else {
+				require.NoError(t, err)
+				assert.EqualValues(t, spec.exp, loaded)
+				assert.EqualValues(t, spec.expPageRes.Total, res.Total)
+				assert.EqualValues(t, spec.expPageRes.NextKey, res.NextKey)
+			}
+
+		})
+	}
+}
```

Comment:
we need to add more tests to cover edge cases we were discussing while reviewing the `Paginate` function.

Reply:
Given https://github.com/regen-network/regen-ledger/pull/154#discussion_r562668406, I believe that what we were discussing was a bit off. But I have added additional tests anyway, please have a look, thanks.